Correctly handle unpacking empty objects

gassmoeller opened the pull request:
Still hunting for the solution to geodynamics/aspect#4984, but at least I think this PR should fix a real issue when we try to unpack an empty object. If the packed buffer `std::vector<char>` is empty, we are not allowed to do anything with its iterators, except for checking for equality between `cbegin` and `cend`. So no matter if the object is trivially copyable or not, we should just return an default constructed object of type T (or crash for the version of the unpack function that unpacks an array, because empty arrays are not allowed).

gassmoeller committed:
Correctly handle packing/unpacking empty objects

masterleinad commented on include/deal.II/base/utilities.h:
```diff
@@ -1252,6 +1252,12 @@ namespace Utilities
       const std::vector<char>::const_iterator &cend,
       std::vector<T> &                         object)
     {
+      // return early if nothing to copy. this avoids
+      // dereferencing iterators if the std::vector<char> is empty,
+      // which would not be allowed.
+      if (cbegin == cend)
+        return;
```
Shouldn't this at least call `object.clear()` before returning?

tjhei commented:
I am a bit confused about this: Is there ever a situation where pack produces an empty buffer? If yes, then unpack needs to handle the reverse operation. Otherwise, passing an empty buffer into pack should be an error.
I tried packing an empty std::vector, it will be nonzero in size and unpack correctly. You can not pack an empty array using pack(), at least I was not able to.

But you are correct that dereferencing begin should be avoided if the buffer is empty.

In other words, I think the bug is somewhere else.

tjhei commented:
The crash happens when unpacking an ``std::vector<std::pair>``, which should never have an empty buffer as a result.

gassmoeller commented:
> I am a bit confused about this: Is there ever a situation where pack produces an empty buffer?

As far as I understand since #13854 we do sometimes send empty messages (empty tuples) that will be unpacked. It should only happen for trivially copyable objects (not the std::vectors in my case). But I think this is still a problem, because all code paths in unpack try to dereference the begin iterator, which is not allowed for empty vectors (memcpy explicitly states  that providing a null pointer for the beginning/end address is undefined behavior, even if the length to copy  is zero).

> In other words, I think the bug is somewhere else.

I agree, even after this fix the consensusalgorithm deadlocks for me because some messages do not seem to arrive. However, this PR at least fixed the exceptions thrown by boost. This is not a fix for the bug, just something that I noticed while looking for the source of the problem.

gassmoeller committed:
clear object before returning

drwells commented on include/deal.II/base/utilities.h:
```diff
@@ -1498,6 +1523,11 @@ namespace Utilities
+    // see if the object is empty. if so, shortcut the whole
+    // operation and return an empty object.
+    if (buffer.size() == 0)
+      return T();
```
This seems wrong - we should throw an exception here like we do below (e.g., what if `T` is not default-constructible?)

bangerth replied:
same here

drwells commented on include/deal.II/base/utilities.h:
```diff
@@ -1425,9 +1443,16 @@ namespace Utilities
+    // dereferencing iterators if the std::vector<char> is empty,
+    // which would not be allowed.
+    if (cbegin == cend)
+      return T();
```
same

bangerth replied:
This should be ok -- we create the to-be-returned object below via
```
  T object;
```
so we already require that the object be default-constructible.

bangerth replied:
But I think that some clang-tidy or other check will ask that you write this instead:
```suggestion
      return {};
```

bangerth reviewed (commented):
I am confused by this as well. `pack()` special-cases the situation where `T=std::tuple<>` (an empty object) to create an empty buffer. But `unpack()` also special-cases this situation, and you shouldn't end up in the place where you fix things. So the question I have is where you are creating an empty buffer. If you only see the problem on the receiving side, you can always put an assertion into the sending side that the message size is nonzero.

I'm not against this patch, but for the moment it looks like you're papering over the problem. I'd love to understand where the problem actually arises.

kronbichler commented:
/rebuild

kronbichler reviewed (approved):
Looks good to me, apart from the comments that @bangerth made.

gassmoeller commented:
> I'm not against this patch, but for the moment it looks like you're papering over the problem. I'd love to understand where the problem actually arises.

I guess you are right. Let me investigate a bit further and pause this PR for now.

bangerth commented:
@gassmoeller Ping?

gassmoeller commented:
I completely forgot about this PR. I rebased and addressed all comments. I am not sure if the additional checks make sense though as pointed out in the discussion above. I can also close the PR.

bangerth commented on include/deal.II/base/utilities.h:
```diff
@@ -1468,6 +1486,13 @@ namespace Utilities
          const std::vector<char>::const_iterator &cend,
          const bool                               allow_compression)
   {
+    // See if the object is empty. If so, shortcut the whole
+    // operation and return an empty object. This avoids
+    // dereferencing iterators if the std::vector<char> is empty,
+    // which would not be allowed.
+    if (cbegin == cend)
+      return {};
```
This does not strike me as right. We should only ever have no data to unpack if `T` has size zero. Can you guard that case here by `Assert (sizeof(T)==0, ExcInternalError())` or some such?

bangerth commented on include/deal.II/base/utilities.h:
```diff
@@ -1285,6 +1285,15 @@ namespace Utilities
       const std::vector<char>::const_iterator &cend,
       std::vector<T>                          &object)
     {
+      // return early if nothing to copy. this avoids
+      // dereferencing iterators if the std::vector<char> is empty,
+      // which would not be allowed.
+      if (cbegin == cend)
+        {
+          object.clear();
+          return;
+        }
```
I would hope that this only ever happens if we had a corresponding case where we're serializing an empty vector by an empty array. But I don't think we do. Rather, I believe we serialize an empty vector by putting a zero integer into the buffer. (We could change this, but that's a separate issue.)

bangerth commented on include/deal.II/base/utilities.h:
```diff
@@ -1334,6 +1343,15 @@ namespace Utilities
       const std::vector<char>::const_iterator &cend,
       std::vector<std::vector<T>>             &object)
     {
+      // return early if nothing to copy. this avoids
+      // dereferencing iterators if the std::vector<char> is empty,
+      // which would not be allowed.
+      if (cbegin == cend)
+        {
+          object.clear();
+          return;
+        }
```
same here

bangerth commented on include/deal.II/base/utilities.h:
```diff
@@ -1537,6 +1562,11 @@ namespace Utilities
   T
   unpack(const std::vector<char> &buffer, const bool allow_compression)
   {
+    // see if the object is empty. if so, shortcut the whole
+    // operation and return an empty object.
+    if (buffer.size() == 0)
+      return {};
```
same as above with the zero-sized object

bangerth commented on include/deal.II/base/utilities.h:
```diff
@@ -1548,6 +1578,15 @@ namespace Utilities
          T (&unpacked_object)[N],
          const bool allow_compression)
   {
+    // empty arrays are not allowed and so unpacking requires to
+    // have a non-zero amount of input data
+    Assert(
+      cbegin != cend,
+      ExcMessage(
+        "Utilities::unpack was asked to unpack "
+        "an array, but no input data to unpack was provided. Empty arrays are "
+        "not allowed."));
```
This is ok, as is the next change. In fact, in all of the places where you currently have an early exit, adding this assertion seems like the right thing to do as well.

bangerth reviewed (changes requested):
I'm not convinced this is right. The `pack()` and `unpack()` functions must match. If `pack()` cannot create empty arrays, then `unpack()` should not accept empty arrays either.

bangerth commented:
@gassmoeller Ping?

gassmoeller commented:
Lets close this PR. It was created while looking for another bug, which turned out to be unrelated to deal.II, and we havent seen any reports that this PR is actually necessary.
